dlib/cuda/cudnn_dlibapi.h

```diff
@@ -4,16 +4,28 @@
 #define DLIB_DNN_CuDNN_H_
 
 #ifdef DLIB_USE_CUDA
-
 #include "cuda_errors.h"
 #include <memory>
 #include "cuda_data_ptr.h"
+#endif // DLIB_USE_CUDA
 
 namespace dlib
 {
     class tensor;
     class resizable_tensor;
 
+    // ----------------------------------------------------------------------------------------
+
+    /*!
+        This enum is used to determine the mode of operation for certain functions
+        (such as gemm and softmax) in Dlib. It specifies whether the calculation
+        should be performed based on the matrix field in nr()xnc() or if the matrix
+        should be considered in num_samples()xk(). This helps in organizing tensor
+        computations more efficiently according to the required dimensions.
+    !*/
+    enum class operation_mode { CHANNEL_WISE = 0, PLANE_WISE = 1 };
```

Comment:
This is nice.  Move it into some other file though since the docs in stuff like `layers_abstract.h` refer to it.  Probably dlib/cuda/tensor.h and dlib/cuda/tensor_abstract.h are the best place.

Reply:
This may seem less "normal" as a location for this enum, but because of the some complex links between the different classes, it's the only place I've found that leads to visibility by all the other classes, whatever the compilation model. Putting it in tensor.h or its abstract class doesn't work, for instance.

Reply:
Ah, yeah I see.  Put it in its own header then and #include it here.  I only care since cudnn_dlibapi.h is a private detail.  But this enum is exposed to users, so it's got to be public somewhere.  So in some header users are meant to read (and not full of stuff they aren't supposed to read in particular).

Reply:
OK, the enum has been moved to its own header, as proposed.